src/java/org/apache/cassandra/service/accord/AccordExecutorAbstractLockLoop.java

```diff
@@ -266,4 +273,18 @@ protected void runWithoutLock(Interruptible.State state) throws InterruptedExcep
             }
         }
     }
+
+    @Override
+    public void shutdown()
+    {
+        shutdown = true;
```

Comment:
this isn't volatile so how do we make sure the visibility is correct for all versions?

Reply:
here is what i see for `AccordExecutorSyncSubmit`

1) `shutdown` is called in random thread during shutdown
2) `notifyWorkExclusive` does `hasWork.signal();`
3) with lock case only checks this lock once we have no work to do, which is after we access this field

so from a JMM point of view this doesn't look like it actually forces the visibility?

Reply:
Also, we don't block *new* work once we shutdown, so if the node is very busy this will keep going?

Reply:
Visibility is enforced by other volatile accesses. The writer notifies work as you correctly surmised (which involves a volatile write), and the readers are all guarded by volatile reads, both on the queues they are consuming and the lock they are guarded by (that we notify of work). 

it is true that we don’t reject new work but this executor *should* be shutdown after networking and we should not really reject any work being produced by the local machine during shutdown.